source/common/network/listener_impl.cc

```diff
@@ -27,7 +27,11 @@ void ListenerImpl::listenCallback(evconnlistener*, evutil_socket_t fd, sockaddr*
                                   int remote_addr_len, void* arg) {
   ListenerImpl* listener = static_cast<ListenerImpl*>(arg);
 
-  Address::InstanceConstSharedPtr final_local_address = listener->socket_.localAddress();
+  // Get the local address from the new socket.  This is different from the listener address, if
+  // the listener is listening on the all hosts address (e.g., 0.0.0.0 for IPv4).
+  Address::InstanceConstSharedPtr local_address = Address::addressFromFd(fd);
```

Comment:
P.S., assuming we do forward with this change, it will need tests as already mentioned. The way to do that is going to be to add some type of factory/override method that yields the "address from fd" that can then be implemented in test code to test various permutations. Same for the proxy proto case also.